Add support for all streams

theViz343 opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
This PR adds support for unsubscribed streams and never_subscribed streams in ZT. It creates new data structures and accessor methods to access them. This PR is needed for further work on reducing stream related issues and adding support for stream (un)subscription.

### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `Adding streams throws key error`
- [ ] Fully fixes #
- [x] Partially fixes issue #816 
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [x] Merge will enable work on #1387 

### How did you test this?
<!-- [x] all that apply -->
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [x] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [x] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->

theViz343 committed:
model/api_types: Add data structures to support non-subscribed streams.

This commit creates data structures, similar to stream_dict, to support
unsubscribed and never-subscribed streams. These streams are available
in the fetched initial data which is used to populate the new data
structures using the _register_non_subscribed_streams function.

theViz343 committed:
model: Add stream id accessor methods.

This commit adds two accessor methods - _get_stream_from_id and
get_all_stream_ids. These two accessor methods help in preparation for
including specific stream property accessor methods in the next set of
commits.

Tests added.

theViz343 committed:
refactor: model: Use get_stream_name instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
stream "name" with the new stream property accessor method
"get_stream_name".

Test added.

theViz343 committed:
refactor: model/views: Use get_stream_subscribers instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"subscribers" with the new stream property accessor method
"get_stream_subscribers".

Test added.

theViz343 committed:
refactor: model: Use stream_date_created (get/set)ter methods.

This commit replaces the usage of directly indexing stream_dict for
"date_created" data with the new stream property accessor methods
"get_stream_date_created" and "set_stream_date_created".

Tests added.

theViz343 committed:
refactor: model: Use is_stream_web_public instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"is_web_public" data with the new stream property accessor method
"is_stream_web_public".

Test added.

theViz343 committed:
refactor: model: Use stream_message_retention_days (get/set)ter methods.

This commit replaces the usage of directly indexing stream_dict for
"message_retention_days" data with the new stream property accessor
methods "get_stream_message_retention_days" and
"set_stream_message_retention_days".

Test added.

theViz343 committed:
refactor: model: Use get_subscription_color instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"color" data with the new stream property accessor method
"get_subscription_color".

Test added.

theViz343 committed:
refactor: model: Use is_stream_invite_only instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"is_invite_only" data with the new stream property accessor method
"is_stream_invite_only".

Test added.

theViz343 committed:
model: Add get_all_subscription_ids accessor method.

This commit adds the subscription ids accessor method
"get_all_stream_ids". This helps in preparation for including
subscription color accessor method added in the next commit.

Tests added.

theViz343 committed:
refactor: model/views: Use get_stream_post_policy accessor method.

This commit replaces the usage of directly indexing stream_dict for
"stream_post_policy" with the new stream property accessor method
"get_stream_post_policy".

Test added.

theViz343 committed:
refactor: model/views: Use is_stream_announcement_only accessor method.

This commit replaces the usage of directly indexing stream_dict for
"is_announcement_only" data with the new stream property accessor method
"is_stream_announcement_only".

Test added.

theViz343 committed:
refactor: model/views: Use is_stream_history_public_to_subscribers method.

This commit replaces the usage of directly indexing stream_dict for
"history_public_to_subscribers" data with the new stream property
accessor method "is_stream_history_public_to_subscribers".

Test added.

theViz343 committed:
refactor: model/views: Use get_stream_weekly_traffic method.

This commit replaces the usage of directly indexing stream_dict for
"stream_weekly_traffic" data with the new stream property
accessor method "get_stream_weekly_traffic".

Test added.

theViz343 committed:
refactor: model/views: Use get_stream_rendered_description method.

This commit replaces the usage of directly indexing stream_dict for
"rendered_description" data with the new stream property
accessor method "get_stream_rendered_description".

Test added.

theViz343 committed:
refactor: model/views: Use get_subscription_email instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"email_address" data with the new stream property accessor method
"get_subscription_email".

Test added.

prah23 commented on zulipterminal/model.py:
```diff
@@ -1243,6 +1243,13 @@ def get_subscription_color(self, stream_id: int) -> Optional[str]:
+            return self.stream_dict[stream_id]["email_address"]
+        elif stream_id in self._unsubscribed_streams:
+            return self._unsubscribed_streams[stream_id]["email_address"]
+        return None
```
Since this method is only being used in the `StreamInfoView` and you haven't added the feature of viewing stream info for a never subscribed stream, can the base condition of returning `None` be avoided? The return value of this method is currently attached to a copy command, and having to copy a `None` to the clipboard doesn't seem appropriate.

theViz343 replied:
I've made it so that the method raises a RuntimeError if the stream is not a subscription (or does not exist).

prah23 reviewed (changes requested):
Thanks for the PR @theViz343!
This is a neat refactor; I've just left some minor points of feedback.

theViz343 committed:
refactor: model: Improve stream_dict typing.

This commit improves the typing of stream_dict by using Subscription
TypedDict instead of Dict[str, Any].

Tests updated.

theViz343 committed:
refactor: model: Rename stream_dict to _subscribed_streams.

This commit renames stream_dict to _subscribed_streams. This change is
necessary due to the introduction of _unsubscribed_streams and
never_subscribed streams in the model.

Tests updated.

Subhasish-Behera commented on tests/model/test_model.py:
```diff
@@ -1779,6 +1779,32 @@ def test_get_all_stream_ids(
+            MODEL + "._get_stream_from_id", return_value=get_stream_from_id_fixture
+        )
+        assert model.get_stream_name(stream_id) == expected_value
+        model._get_stream_from_id.assert_called_once()
```
can `assert_called_once_with` be used here?

Subhasish-Behera commented:
@theViz343  Thanks for working on this! This seems to work well and was pretty easy to understand because of the commit structure and commit messages.
Apart from that, I have some minor code suggestions in the comments.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1215,6 +1215,9 @@ def set_stream_message_retention_days(
     ) -> None:
         self._get_stream_from_id(stream_id)["message_retention_days"] = value
 
+    def is_stream_invite_only(self, stream_id: int) -> bool:
+        return self._get_stream_from_id(stream_id)["invite_only"]
```
Similarly to is_web_public, we already have a method for this, which you've updated to use this accessor.

There is no use of this data outside of the model in any case, other than through stream_access_type(), so there is no motivation for introducing this accessor.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1292,8 +1292,10 @@ def __init__(self, controller: Any, stream_id: int) -> None:
+                controller.model.get_stream_post_policy(self.stream_id)
+            ]
         else:
             if stream.get("is_announcement_only"):
```
This existing UI code incorporates implicit checking of the server version, which would be better wrapped into the model under `get_stream_post_policy`, rather than the two functions introduced in this commit and the next.

neiljp commented on zulipterminal/model.py:
```diff
@@ -174,7 +174,7 @@ def __init__(self, controller: Any) -> None:
         self.users = self.get_all_users()
 
-        self.stream_dict: Dict[int, Any] = {}
+        self.stream_dict: Dict[int, Subscription] = {}
```
This change, and possibly others in this commit seem much more related to those earlier in the PR than here at the end?

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1268,7 +1268,7 @@ class StreamInfoView(PopUpView):
         self.stream_id = stream_id
         self.controller = controller
-        stream = controller.model.stream_dict[stream_id]
+        stream = controller.model._subscribed_streams[stream_id]
```
This suggests we haven't covered accessing the stream or subscription properties sufficiently?

neiljp reviewed (commented):
@theViz343 This has grown quite a lot since I last looked! I started a review of this before Hari, so apologies if this overlaps a little. The end result is great, though I think we can shrink it a little.

The PR seems to be split into two parts:
- adding the un/never-subscribed data, improving the data types, and generalizing the current code to avoid eg. the KeyErrors
- adding the encapsulation via the accessors, followed by the rename

Both of these parts have value, but I'd suggest grouping the first set together at the start so that we can consider them together, and so potentially merge the earlier commits in this PR first.

There are a lot of commits, particularly for the encapsulating functions, and isolating adding different methods and tests per commit is great. However, some of these functions are not used outside the model, and don't appear to be preparing for anything, or the effect is replicated in other functions already. You also have some 'setters', which we likely don't need for users of the model.

With a smaller set of new methods I think the changes will be clearer, but it may be worth discussing whether adding so many functions is the way to go, or rather using more general stream_properties and subscription_properties functions?

theViz343 commented on zulipterminal/api_types.py:
```diff
@@ -197,37 +197,26 @@ class Message(TypedDict, total=False):
     description: str
     rendered_description: str
-    date_created: int  # NOTE: new in Zulip 4.0 / ZFL 30
+    date_created: Optional[int]  # NOTE: new in Zulip 4.0 / ZFL 30
```
Yes, the part where we add the `date_created` key and set it to None fails the typecheck. Since we get the data from the server initially, mypy doesn't catch that `date_created` is not present. We could account for this by using `NotRequired[Optional[int]]` instead, but that would purely be for documentation since it passes the typecheck with just `Optional[int]`. I'll also add a comment.

theViz343 commented on zulipterminal/model.py:
```diff
@@ -1176,6 +1176,22 @@ def _register_non_subscribed_streams(
+        else:
+            raise RuntimeError(f"Stream with id {stream_id} does not exist!")
+
+    def get_all_stream_ids(self) -> List[int]:
```
Yeah, you're right, both are helper functions and are used internally only. I'll change the commit description to indicate that, and also maybe change the name for the `get_all_stream_ids` function to `_get_all_stream_ids`.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1156,6 +1164,18 @@ def user_name_from_id(self, user_id: int) -> str:
+        never_subscribed_streams: List[Stream],
+    ) -> None:
+        self._unsubscribed_streams = {
+            stream["stream_id"]: stream for stream in unsubscribed_streams
```
Resolved in #1419 :+1:

neiljp commented on zulipterminal/model.py:
```diff
@@ -174,11 +175,18 @@ def __init__(self, controller: Any) -> None:
         self.users = self.get_all_users()
 
         self.stream_dict: Dict[int, Any] = {}
```
Resolved in #1419 :+1:
